README.md
[![Build](https://github.com/Smashing/smashing/actions/workflows/main.yml/badge.svg)](https://github.com/Smashing/smashing/actions/workflows/main.yml)
[![Gitter chat](https://badges.gitter.im/smashing.svg)](https://gitter.im/Smashing)

# [Smashing](https://github.com/Smashing/smashing/wiki)

Smashing, the spiritual successor to [Dashing](https://github.com/Shopify/dashing), is a Sinatra based framework that lets you build excellent dashboards. It looks especially great on TVs.

![Smashing logo](./.github/docs/assets/screenshot.png)

## Community

Feel free to submit issues for bugs, new features, and enhancements in [GitHub](https://github.com/Smashing/smashing/issues). For more general questions, or help with widgets, please use the [gitter chatroom](https://gitter.im/Smashing).

## Installation

```bash
# Install bundler
$ gem install bundler
# Install smashing
$ gem install smashing
# Create a new project
$ smashing new my-project
# Change Directory into the project
$ cd my-project
# Install the bundle of project specific gems
$ bundle
# Start the example dashboard!
$ smashing start
```

[Check out our wiki](https://github.com/Smashing/smashing/wiki).

Note: This is a fork of the Dashing project, which is no longer being maintained. Read about that [here](https://github.com/Shopify/dashing/issues/711).

## License
Distributed under the [MIT license](MIT-LICENSE).


CHANGELOG.md
# Changelog

## 2023-??-?? - 1.3.7

- [199](https://github.com/Smashing/smashing/pull/199)  Bump rack from 2.2.4 to 2.2.6.3

## 2022-12-08 - 1.3.6

- [190](https://github.com/Smashing/smashing/pull/190) Add sprockets precompile rake task
- [196](https://github.com/Smashing/smashing/pull/196) Replaced open with URI.open, as Kernel#open was deprecated in Ruby 2.7 and doens't work anymore with Ruby 3 (thanks @thomaswitt)
- [197](https://github.com/Smashing/smashing/pull/197) Update rack requirement from ~> 2.2 to >= 2.2, < 4.0 (thanks to @dependabot)

## 2021-07-01 - 1.3.5

- [186](https://github.com/Smashing/smashing/pull/186) Escape widget error message (thanks @gebhardtr) ([CVE-2021-35440](https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2021-35440))
- [142](https://github.com/Smashing/smashing/issues/142) Add screenshots
- [166](https://github.com/Smashing/smashing/issues/166) Update installation instructions docs for Windows
- [168](https://github.com/Smashing/smashing/issues/168) Update installation instructions docs for Windows
- [174](https://github.com/Smashing/smashing/issues/174) Update troubleshooting docs with note about thin_http_parser error
- [167](https://github.com/Smashing/smashing/issues/167) Update troubleshooting docs with note about raspberry pi error

## 2021-03-07 - 1.3.4

- [184](https://github.com/Smashing/smashing/pull/184) Add switch to not filter events (thanks @toy)

## 2021-03-06 - 1.3.3

- [182](https://github.com/Smashing/smashing/pull/182) Create EventSource at the right moment (thanks @toy)

## 2021-02-27 - 1.3.2

- [179](https://github.com/Smashing/smashing/pull/179) Replace Travis CI by GH actions
- [178](https://github.com/Smashing/smashing/pull/178) Relax dependencies; pinned minor, free patch version (thanks @dylanratcliffe)
- [173](https://github.com/Smashing/smashing/pull/173) Send only data that is needed on the dashboard (thanks @toy)

## 2021-01-30 - 1.3.1

- [177](https://github.com/Smashing/smashing/pull/177) Remove websocket connections that have been closed in send_event (thanks @ssgelm)

## 2020-06-04 - 1.3.0

- [#157](https://github.com/Smashing/smashing/pull/157) Update gems
- [#155](https://github.com/Smashing/smashing/pull/155) Fix bug causing AnimatedValue to freeze after same value is entered
- [#162](https://github.com/Smashing/smashing/pull/162) Add version command
- [#163](https://github.com/Smashing/smashing/pull/163) Disable the twitter job and comment out the gem dependency
- Dropped Ruby 2.2 in Travis due to rack requiring >= 2.3.0

## 2020-05-19 - 1.2.0

- [#151](https://github.com/Smashing/smashing/pull/151) Upgrading font awesome to 5.11
- [#153](https://github.com/Smashing/smashing/pull/153) Fix GitHub security alert updating to rake 12.3.3
- [#147](https://github.com/Smashing/smashing/pull/147) Checking for existing displayedValue
- [#141](https://github.com/Smashing/smashing/pull/141) Disable warnings in test, circular require warning we're clouding up
- [#135](https://github.com/Smashing/smashing/pull/135) Fixing favicon location
- [#137](https://github.com/Smashing/smashing/pull/137) Use {} as fallback for history.yaml
- [#103](https://github.com/Smashing/smashing/pull/103) Don't display exception pages to users by default
- [#107](https://github.com/Smashing/smashing/pull/107) Update to latest gridster to fix touch input bug
- [#95](https://github.com/Smashing/smashing/pull/95) Silence rake warnings about unused vars
- [#96](https://github.com/Smashing/smashing/pull/96) Added Batman filter 'shortenedBytes'

## 2017-12-25 - 1.1.0

- [#25](https://github.com/Smashing/smashing/pull/25) Add data transformation method to widget base
- [#27](https://github.com/Smashing/smashing/pull/27) Add error string when widget not rendered
- [73e75f](https://github.com/Smashing/smashing/commit/73e75f) Switched background colors for 'danger' and 'warning'
- [#53](https://github.com/Smashing/smashing/pull/52) Use old-style hash syntax to enable compatibility with ruby 1.8.7
- [7a6d67](https://github.com/Smashing/smashing/commit/7a6d67) Update gemspec to have a valid link to the Smashing GitHub pages site
- [#56](https://github.com/Smashing/smashing/pull/56) Change update time on events
- [#60](https://github.com/Smashing/smashing/pull/60) Fix favicon url
- [#11](https://github.com/Smashing/smashing/pull/11) Update Rufus-Scheduler version
- [#67](https://github.com/Smashing/smashing/pull/67) Capture Interrupt error (Ctrl+c) and cleanly exit
- [fa7cf4](https://github.com/Smashing/smashing/commit/fa7cf4) Don't multiply the updatedAt timestamp by 1000
- [855348](https://github.com/Smashing/smashing/commit/855348) Travis: Change to pre-installed Rubies, drop 2.1
- [4e2b7f](https://github.com/Smashing/smashing/commit/4e2b7f) Update all gems
- [305193](https://github.com/Smashing/smashing/commit/305193) Facilitate testing

## 2016-10-23 - 1.0.0

- First release after Dashing: https://github.com/Shopify/dashing/issues/711


test/app_test.rb
require 'test_helper'
require 'haml'

class AppTest < Dashing::Test
  def setup
    @connection = []
    # stop sinatra from handling Hash value specially which makes it merge new value into previous one
    app.settings.connections = nil
    app.settings.connections = {@connection => ['some_widget']}
    app.settings.auth_token = nil
    app.settings.default_dashboard = nil
    app.settings.history_file = File.join(Dir.tmpdir, 'history.yml')
    app.settings.history = nil
    app.settings.history = {}
  end

  def test_redirect_to_first_dashboard
    with_generated_project do
      get '/'
      assert_equal 302, last_response.status
      assert_equal 'http://example.org/sample', last_response.location
    end
  end

  def test_redirect_to_first_dashboard_without_erb
    with_generated_project do |dir|
      FileUtils.touch(File.join(dir, "dashboards/htmltest.html"))
      get '/'
      assert_equal 302, last_response.status
      assert_equal 'http://example.org/htmltest', last_response.location
    end
  end

  def test_redirect_to_default_dashboard
    with_generated_project do
      app.settings.default_dashboard = 'test1'
      get '/'
      assert_equal 302, last_response.status
      assert_equal 'http://example.org/test1', last_response.location
    end
  end

  def test_errors_out_when_no_dashboards_available
    with_generated_project do
      app.settings.views = File.join(app.settings.root, 'lib')

      get '/'
      assert_equal 500, last_response.status
    end
  end

  def test_post_widgets_without_auth_token
    post '/widgets/some_widget', JSON.generate({value: 6})
    assert_equal 204, last_response.status

    assert_equal 1, @connection.length
    data = parse_data @connection[0]
    assert_equal 6, data['value']
    assert_equal 'some_widget', data['id']
    assert data['updatedAt']
  end

  def test_post_widgets_with_invalid_auth_token
    app.settings.auth_token = 'sekrit'
    post '/widgets/some_widget', JSON.generate({value: 9})
    assert_equal 401, last_response.status
  end

  def test_post_widgets_with_valid_auth_token
    app.settings.auth_token = 'sekrit'
    post '/widgets/some_widget', JSON.generate({value: 9, auth_token: 'sekrit'})
    assert_equal 204, last_response.status
  end

  def test_get_events
    post '/widgets/some_widget', JSON.generate({value: 8})
    assert_equal 204, last_response.status

    get '/events'
    assert_equal 200, last_response.status
    assert_equal 8, parse_data(@connection[0])['value']
  end

  # rubocop:disable Metrics/AbcSize
  # rubocop:disable Metrics/MethodLength
  def test_get_events_by_id
    with_generated_project do
      post '/widgets/synergy', JSON.generate({ value: 3 })
      assert_equal 204, last_response.status

      post '/widgets/valuation', JSON.generate({ value: 'a' })
      assert_equal 204, last_response.status

      get '/events?ids=synergy%2Cvaluation'
      assert_equal 200, last_response.status

      events = last_response.body.split(/\n+/) # two events, printed with blank lines separating them
      assert_equal 2, events.length

      assert_equal ['synergy', 3], parse_data(events[0]).values_at('id', 'value')
      assert_equal %w[valuation a], parse_data(events[1]).values_at('id', 'value')
    end
  end

  def test_get_events_by_unknown_id
    with_generated_project do
      post '/widgets/synergy', JSON.generate({ value: 3 })
      assert_equal 204, last_response.status

      get '/events?ids=unknown'
      assert_equal 200, last_response.status
      assert_equal '', last_response.body
    end
  end

  def test_get_events_without_filter
    with_generated_project do
      post '/widgets/foobar', JSON.generate({ value: 'x' })
      assert_equal 204, last_response.status

      get '/events'
      assert_equal 200, last_response.status

      events = last_response.body.split(/\n+/)
      assert_equal 1, events.length

      assert_equal ['foobar', 'x'], parse_data(events[0]).values_at('id', 'value')
    end
  end

  def test_dashboard_events
    post '/dashboards/my_super_sweet_dashboard', JSON.generate({event: 'reload'})
    assert_equal 204, last_response.status

    get '/events'
    assert_equal 200, last_response.status
    assert_equal 'dashboards', parse_event(@connection[0])
    assert_equal 'reload', parse_data(@connection[0])['event']
  end

  def test_get_dashboard
    with_generated_project do
      get '/sampletv'
      assert_equal 200, last_response.status
      assert_includes last_response.body, 'class="grid"'
      assert_includes last_response.body, "DOCTYPE"
    end
  end

  def test_page_title_set_correctly
    with_generated_project do
      get '/sampletv'
      assert_includes last_response.body, '<title>1080p dashboard</title>'
    end
  end

  def test_get_haml_dashboard
    app.template_languages << :haml

    with_generated_project do |dir|
      File.write(File.join(dir, 'dashboards/hamltest.haml'), '.gridster')
      get '/hamltest'
      assert_equal 200, last_response.status
      assert_includes last_response.body, "class='gridster'"
    end
  end

  def test_get_haml_widget
    app.template_languages << :haml

    with_generated_project do |dir|
      File.write(File.join(dir, 'widgets/clock/clock.haml'), '%h1 haml')
      File.unlink(File.join(dir, 'widgets/clock/clock.html'))
      get '/views/clock.html'
      assert_equal 200, last_response.status
      assert_includes last_response.body, '<h1>haml</h1>'
    end
  end

  def test_get_nonexistent_dashboard
    with_generated_project do
      get '/nodashboard'
      assert_equal 404, last_response.status
    end
  end

  def test_get_widget
    with_generated_project do
      get '/views/meter.html'
      assert_equal 200, last_response.status
      assert_includes last_response.body, 'class="meter"'
    end
  end

  def test_get_nonexistent_widget
    with_generated_project do
      get '/views/nowidget.html'
      assert_equal 200, last_response.status
      assert_equal last_response.body, 'Drats! Unable to find a widget file named: nowidget to render.'
    end
  end

  def test_get_xss_widget
    with_generated_project do
      get CGI.escape('/views/nowidget-<h1>.html')
      assert_equal 200, last_response.status
      assert_equal last_response.body, 'Drats! Unable to find a widget file named: nowidget-&lt;h1&gt; to render.'
    end
  end

  def with_generated_project
    source_path = File.expand_path('../../templates', __FILE__)

    temp do |dir|
      cli = Dashing::CLI.new
      cli.stubs(:source_paths).returns([source_path])
      silent { cli.new 'new_project' }

      app.settings.public_folder = File.join(dir, 'new_project/public')
      app.settings.views = File.join(dir, 'new_project/dashboards')
      app.settings.root = File.join(dir, 'new_project')

      app.settings.raise_errors = true
      app.settings.dump_errors = false
      app.settings.show_exceptions = false

      yield app.settings.root
    end
  end

  def app
    Sinatra::Application
  end

  def parse_data(string)
    JSON.parse string[/data: (.+)/, 1]
  end

  def parse_event(string)
    string[/event: (.+)/, 1]
  end
end


test/cli_test.rb
require 'test_helper'

class CLITest < Dashing::Test
  def setup
    @cli = Dashing::CLI.new
  end

  def test_version_command
    version = Dashing::VERSION
    output, _ = capture_io do
      @cli.__print_version
    end

    assert_includes output, "#{version}"
  end

  def test_new_task_creates_project_directory
    app_name = 'custom_dashboard'
    @cli.stubs(:directory).with(:project, app_name).once
    @cli.new(app_name)
  end

  def test_generate_task_delegates_to_type
    types = %w(widget dashboard job)

    types.each do |type|
      @cli.stubs(:public_send).with("generate_#{type}".to_sym, 'name').once
      @cli.generate(type, 'name')
    end
  end

  def test_generate_task_warns_when_generator_is_not_defined
    output, _ = capture_io do
      @cli.generate('wtf', 'name')
    end

    assert_includes output, 'Invalid generator'
  end

  def test_generate_widget_creates_a_new_widget
    @cli.stubs(:directory).with(:widget, 'widgets').once
    @cli.generate_widget('WidgetName')
    assert_equal 'widget_name', @cli.name
  end

  def test_generate_dashboard_creates_a_new_dashboard
    @cli.stubs(:directory).with(:dashboard, 'dashboards').once
    @cli.generate_dashboard('DashBoardName')
    assert_equal 'dash_board_name', @cli.name
  end

  def test_generate_job_creates_a_new_job
    @cli.stubs(:directory).with(:job, 'jobs').once
    @cli.generate_job('MyCustomJob')
    assert_equal 'my_custom_job', @cli.name
  end

  def test_install_task_requests_gist_from_downloader
    return_value = { 'files' => [] }
    Dashing::Downloader.stubs(:get_gist).with(123).returns(return_value).once

    capture_io { @cli.install(123) }
  end

  def test_install_task_calls_create_file_for_each_valid_file_in_gist
    json_response = <<-JSON
      {
        "files": {
          "ruby_job.rb": { "content": "some job content" },
          "num.html": { "content": "some html content" },
          "num.scss": { "content": "some sass content" },
          "num.coffee": { "content": "some coffee content" }
        }
      }
    JSON

    Dir.stubs(:pwd).returns('')

    Dashing::Downloader.stubs(:get_gist).returns(JSON.parse(json_response))
    @cli.stubs(:create_file).with('/jobs/ruby_job.rb', 'some job content', {:skip => false}).once
    @cli.stubs(:create_file).with('/widgets/num/num.html', 'some html content', {:skip => false}).once
    @cli.stubs(:create_file).with('/widgets/num/num.scss', 'some sass content', {:skip => false}).once
    @cli.stubs(:create_file).with('/widgets/num/num.coffee', 'some coffee content', {:skip => false}).once

    capture_io { @cli.install(123) }
  end

  def test_install_task_ignores_invalid_files
    json_response = <<-JSON
      {
        "files": {
          "ruby_job.js": { "content": "some job content" },
          "num.css": { "content": "some sass content" }
        }
      }
    JSON

    Dashing::Downloader.stubs(:get_gist).returns(JSON.parse(json_response))
    @cli.stubs(:create_file).never

    capture_io { @cli.install(123) }
  end

  def test_install_task_warns_when_gist_not_found
    error = OpenURI::HTTPError.new('error', mock())
    Dashing::Downloader.stubs(:get_gist).raises(error)

    output, _ = capture_io { @cli.install(123) }

    assert_includes output, 'Could not find gist at '
  end

  def test_start_task_starts_thin_with_default_port
    command = 'bundle exec thin -R config.ru start -p 3030 '
    @cli.stubs(:run_command).with(command).once
    @cli.start
  end

  def test_start_task_starts_thin_with_specified_port
    command = 'bundle exec thin -R config.ru start -p 2020'
    @cli.stubs(:run_command).with(command).once
    @cli.start('-p', '2020')
  end

  def test_start_task_supports_job_path_option
    commands = [
      'export JOB_PATH=other_spot; ',
      'bundle exec thin -R config.ru start -p 3030 '
    ]

    @cli.stubs(:options).returns(job_path: 'other_spot')
    @cli.stubs(:run_command).with(commands.join('')).once
    @cli.start
  end

  def test_stop_task_stops_thin_server
    @cli.stubs(:run_command).with('bundle exec thin stop')
    @cli.stop
  end

  def test_job_task_requires_job_file
    Dir.stubs(:pwd).returns('')
    @cli.stubs(:require_file).with('/jobs/special_job.rb').once

    @cli.job('special_job')
  end

  def test_job_task_requires_every_ruby_file_in_lib
    Dir.stubs(:pwd).returns('')
    Dir.stubs(:[]).returns(['lib/dashing/cli.rb', 'lib/dashing.rb'])
    @cli.stubs(:require_file).times(3)

    @cli.job('special_job')
  end

  def test_job_sets_auth_token
    @cli.class.stubs(:auth_token=).with('my_token').once
    @cli.stubs(:require_file)

    @cli.job('my_job', 'my_token')
  end

  def test_hyphenate_lowers_and_hyphenates_inputs
    assertion_map = {
      'Power' => 'power',
      'POWER' => 'power',
      'PowerRangers' => 'power-rangers',
      'Power_ranger' => 'power-ranger',
      'SuperPowerRangers' => 'super-power-rangers'
    }

    assertion_map.each do |input, expected|
      assert_equal expected, Dashing::CLI.hyphenate(input)
    end
  end

end


test/test_helper.rb
require 'simplecov'
SimpleCov.start do
  add_filter "/vendor/"
  add_filter "/test/"
end

require 'rack/test'
require 'stringio'
require 'tmpdir'
require 'fakeweb'
require 'minitest/autorun'
require 'minitest/pride'
require 'mocha/setup'

require_relative '../lib/dashing'

FakeWeb.allow_net_connect = false

ENV['RACK_ENV'] = 'test'
WORKING_DIRECTORY = Dir.pwd.freeze
ARGV.clear

def load_quietly(file)
  Minitest::Test.new(nil).capture_io do
    load file
  end
end

def temp
  path = File.expand_path "#{Dir.tmpdir}/#{Time.now.to_i}#{rand(1000)}/"
  FileUtils.mkdir_p path
  Dir.chdir path
  yield path
ensure
  Dir.chdir WORKING_DIRECTORY
  FileUtils.rm_rf(path) if File.exist?(path)
end

module Dashing
  class Test < Minitest::Test
    include Rack::Test::Methods

    alias_method :silent, :capture_io

    def teardown
      FileUtils.rm_f('history.yml')
    end
  end
end


lib/dashing/cli.rb
require 'thor'
require 'open-uri'
require 'dashing/version'

module Dashing
  class CLI < Thor
    include Thor::Actions

    attr_reader :name

    class << self
      attr_accessor :auth_token

      def hyphenate(str)
        return str.downcase if str =~ /^[A-Z-]+$/
        str.gsub('_', '-').gsub(/\B[A-Z]/, '-\&').squeeze('-').downcase
      end
    end

    no_tasks do
      %w(widget dashboard job).each do |type|
        define_method "generate_#{type}" do |name|
          @name = Thor::Util.snake_case(name)
          directory(type.to_sym, "#{type}s")
        end
      end
    end

    desc "--version, -v", "Prints the version"
    map %w[--version -v] => :__print_version
    def __print_version
      say "#{Dashing::VERSION}"
    end

    desc "new PROJECT_NAME", "Sets up ALL THE THINGS needed for your dashboard project."
    def new(name)
      @name = Thor::Util.snake_case(name)
      directory(:project, @name)
    end

    desc "generate (widget/dashboard/job) NAME", "Creates a new widget, dashboard, or job."
    def generate(type, name)
      public_send("generate_#{type}".to_sym, name)
    rescue NoMethodError => _e
      puts "Invalid generator. Either use widget, dashboard, or job"
    end

    desc "install GIST_ID [--skip]", "Installs a new widget from a gist (skip overwrite)."
    def install(gist_id, *args)
      gist = Downloader.get_gist(gist_id)
      public_url = "https://gist.github.com/#{gist_id}"

      install_widget_from_gist(gist, args.include?('--skip'))

      print set_color("Don't forget to edit the ", :yellow)
      print set_color("Gemfile ", :yellow, :bold)
      print set_color("and run ", :yellow)
      print set_color("bundle install ", :yellow, :bold)
      say set_color("if needed. More information for this widget can be found at #{public_url}", :yellow)
    rescue OpenURI::HTTPError => _http_error
      say set_color("Could not find gist at #{public_url}"), :red
    end

    desc "start", "Starts the server in style!"
    method_option :job_path, :desc => "Specify the directory where jobs are stored"
    def start(*args)
      port_option = args.include?('-p') ? '' : ' -p 3030'
      args = args.join(' ')
      command = "bundle exec thin -R config.ru start#{port_option} #{args}"
      command.prepend "export JOB_PATH=#{options[:job_path]}; " if options[:job_path]
      run_command(command)
    end

    desc "stop", "Stops the thin server"
    def stop
      command = "bundle exec thin stop"
      run_command(command)
    end

    desc "job JOB_NAME AUTH_TOKEN(optional)", "Runs the specified job. Make sure to supply your auth token if you have one set."
    def job(name, auth_token = "")
      Dir[File.join(Dir.pwd, 'lib/**/*.rb')].each {|file| require_file(file) }
      self.class.auth_token = auth_token
      f = File.join(Dir.pwd, "jobs", "#{name}.rb")
      require_file(f)
    end

    # map some commands
    map 'g' => :generate
    map 'i' => :install
    map 's' => :start

    private

    def run_command(command)
      begin
        system(command)
      rescue Interrupt => _e
        say "Exiting..."
      end
    end

    def install_widget_from_gist(gist, skip_overwrite)
      gist['files'].each do |file, details|
        if file =~ /\.(html|coffee|scss)\z/
          widget_name = File.basename(file, '.*')
          new_path = File.join(Dir.pwd, 'widgets', widget_name, file)
          create_file(new_path, details['content'], :skip => skip_overwrite)
        elsif file.end_with?('.rb')
          new_path = File.join(Dir.pwd, 'jobs', file)
          create_file(new_path, details['content'], :skip => skip_overwrite)
        end
      end
    end

    def require_file(file)
      require file
    end
  end
end


templates/project/jobs/twitter.rb
require 'twitter'


#### Get your twitter keys & secrets:
#### https://dev.twitter.com/docs/auth/tokens-devtwittercom
twitter = Twitter::REST::Client.new do |config|
  config.consumer_key = 'YOUR_CONSUMER_KEY'
  config.consumer_secret = 'YOUR_CONSUMER_SECRET'
  config.access_token = 'YOUR_OAUTH_TOKEN'
  config.access_token_secret = 'YOUR_OAUTH_SECRET'
end

search_term = URI::encode('#todayilearned')

SCHEDULER.every '10m', first_in: 0 do |job|
  begin
    tweets = twitter.search("#{search_term}")

    if tweets
      tweets = tweets.map do |tweet|
        { name: tweet.user.name, body: tweet.text, avatar: tweet.user.profile_image_url_https }
      end
      send_event('twitter_mentions', comments: tweets)
    end
  rescue Twitter::Error
    puts "\e[33mFor the twitter widget to work, you need to put in your twitter API keys in the jobs/twitter.rb file.\e[0m"
  end
end


lib/dashing/app.rb
require 'sinatra'
require 'sprockets'
require 'sinatra/content_for'
require 'rufus/scheduler'
require 'coffee-script'
require 'sass' if RUBY_VERSION < "2.5.0"
require 'sassc' if RUBY_VERSION >= "2.5.0"
require 'json'
require 'yaml'
require 'thin'

SCHEDULER = Rufus::Scheduler.new

def development?
  ENV['RACK_ENV'] == 'development'
end

def production?
  ENV['RACK_ENV'] == 'production'
end

helpers Sinatra::ContentFor
helpers do
  def protected!
    # override with auth logic
  end

  def h(text)
    Rack::Utils.escape_html(text)
  end

  def authenticated?(token)
    return true unless settings.auth_token
    token && Rack::Utils.secure_compare(settings.auth_token, token)
  end
end

set :root, Dir.pwd
set :sprockets,     Sprockets::Environment.new(settings.root)
set :assets_prefix, '/assets'
set :digest_assets, false
set :server, 'thin'
set :connections, {}
set :history_file, 'history.yml'
set :public_folder, File.join(settings.root, 'public')
set :views, File.join(settings.root, 'dashboards')
set :default_dashboard, nil
set :auth_token, nil
set :template_languages, %i[html erb]

if File.exist?(settings.history_file)
  set :history, YAML.load_file(settings.history_file, fallback: {})
else
  set :history, {}
end

%w(javascripts stylesheets fonts images).each do |path|
  settings.sprockets.append_path("assets/#{path}")
end

['widgets', File.expand_path('../../../javascripts', __FILE__)].each do |path|
  settings.sprockets.append_path(path)
end

not_found do
  send_file File.join(settings.public_folder, '404.html'), :status => 404
end

at_exit do
  File.write(settings.history_file, settings.history.to_yaml)
end

get '/' do
  protected!
  dashboard = settings.default_dashboard || first_dashboard
  raise Exception.new('There are no dashboards available') if not dashboard

  redirect "/" + dashboard
end

get '/events', :provides => 'text/event-stream' do
  protected!
  response.headers['X-Accel-Buffering'] = 'no' # Disable buffering for nginx
  ids = params[:ids] ? params[:ids].split(',').to_set : nil
  stream :keep_open do |out|
    settings.connections[out] = ids
    settings.history.each do |id, event|
      out << event if ids.nil? || ids.include?(id)
    end
    out.callback { settings.connections.delete(out) }
  end
end

get '/:dashboard' do
  protected!
  settings.template_languages.each do |language|
    file = File.join(settings.views, "#{params[:dashboard]}.#{language}")
    return render(language, params[:dashboard].to_sym) if File.exist?(file)
  end

  halt 404
end

post '/dashboards/:id' do
  request.body.rewind
  body = JSON.parse(request.body.read)
  body['dashboard'] ||= params['id']
  if authenticated?(body.delete("auth_token"))
    send_event(params['id'], body, 'dashboards')
    204 # response without entity body
  else
    status 401
    "Invalid API key\n"
  end
end

post '/widgets/:id' do
  request.body.rewind
  body = JSON.parse(request.body.read)
  if authenticated?(body.delete("auth_token"))
    send_event(params['id'], body)
    204 # response without entity body
  else
    status 401
    "Invalid API key\n"
  end
end

get '/views/:widget?.html' do
  protected!
  settings.template_languages.each do |language|
    file = File.join(settings.root, "widgets", params[:widget], "#{params[:widget]}.#{language}")
    return Tilt[language].new(file).render if File.exist?(file)
  end

  "Drats! Unable to find a widget file named: #{h(params[:widget])} to render."
end

Thin::Server.class_eval do
  def stop_with_connection_closing
    Sinatra::Application.settings.connections.dup.each_key(&:close)
    stop_without_connection_closing
  end

  alias_method :stop_without_connection_closing, :stop
  alias_method :stop, :stop_with_connection_closing
end

def send_event(id, body, target=nil)
  body[:id] = id
  body[:updatedAt] ||= (Time.now.to_f * 1000.0).to_i
  event = format_event(body.to_json, target)
  Sinatra::Application.settings.history[id] = event unless target == 'dashboards'
  Sinatra::Application.settings.connections.each { |out, ids|
    begin
      out << event if target == 'dashboards' || ids.nil? || ids.include?(id)
    rescue IOError => e # if the socket is closed an IOError is thrown
      Sinatra::Application.settings.connections.delete(out)
    end
  }
end

def format_event(body, name=nil)
  str = ""
  str << "event: #{name}\n" if name
  str << "data: #{body}\n\n"
end

def first_dashboard
  files = Dir[File.join(settings.views, '*')].collect { |f| File.basename(f, '.*') }
  files -= ['layout']
  files.sort.first
end

def require_glob(relative_glob)
  Dir[File.join(settings.root, relative_glob)].each do |file|
    require file
  end
end

settings_file = File.join(settings.root, 'config/settings.rb')
require settings_file if File.exist?(settings_file)

{}.to_json # Forces your json codec to initialize (in the event that it is lazily loaded). Does this before job threads start.
job_path = ENV["JOB_PATH"] || 'jobs'
require_glob(File.join('lib', '**', '*.rb'))
require_glob(File.join(job_path, '**', '*.rb'))
